agent/agent.c

```diff
@@ -20,23 +20,199 @@
 
 #include "agent.h"
 
+#include <hwangsae/dbus/manager-generated.h>
+#include <hwangsae/dbus/edge-interface-generated.h>
+#include <chamge/chamge.h>
+
+#define DEFAULT_HUB_UID        "abc-987-123"
+#define DEFAULT_BACKEND         CHAMGE_BACKEND_AMQP
+
 struct _HwangsaeAgent
 {
   GApplication parent;
+  Hwangsae1DBusManager *manager;
+  Hwangsae1DBusEdgeInterface *edge_interface;
+  ChamgeHub *chamge_hub;
 };
 
 /* *INDENT-OFF* */
 G_DEFINE_TYPE (HwangsaeAgent, hwangsae_agent, G_TYPE_APPLICATION)
 /* *INDENT-ON* */
 
+static gboolean
+hwangsae_agent_dbus_register (GApplication * app,
+    GDBusConnection * connection, const gchar * object_path, GError ** error)
+{
+  gboolean ret;
+  HwangsaeAgent *self = HWANGSAE_AGENT (app);
+
+  g_debug ("hwangsae_agent_dbus_register");
+
+  g_application_hold (app);
+
+  /* chain up */
+  ret =
+      G_APPLICATION_CLASS (hwangsae_agent_parent_class)->dbus_register
+      (app, connection, object_path, error);
+
+  if (ret &&
+      !g_dbus_interface_skeleton_export (G_DBUS_INTERFACE_SKELETON
+          (self->manager), connection,
+          "/org/hwangsaeul/Hwangsaeul1/Manager", error)) {
+    g_warning ("Failed to export Chamge1 D-Bus interface (reason: %s)",
+        (*error)->message);
+  }
+
+  if (ret &&
+      !g_dbus_interface_skeleton_export (G_DBUS_INTERFACE_SKELETON
+          (self->edge_interface), connection,
+          "/org/hwangsaeul/Hwangsaeul1/EdgeInterface", error)) {
+    g_warning ("Failed to export Chamge1 D-Bus interface (reason: %s)",
+        (*error)->message);
+  }
+
+  return ret;
+}
+
+static void
+hwangsae_agent_dbus_unregister (GApplication * app,
+    GDBusConnection * connection, const gchar * object_path)
+{
+  HwangsaeAgent *self = HWANGSAE_AGENT (app);
+
+  g_debug ("hwangsae_agent_dbus_unregister");
+
+  if (self->manager)
+    g_dbus_interface_skeleton_unexport (G_DBUS_INTERFACE_SKELETON
+        (self->manager));
+
+  if (self->edge_interface)
+    g_dbus_interface_skeleton_unexport (G_DBUS_INTERFACE_SKELETON
+        (self->edge_interface));
+
+  g_application_release (app);
+
+  /* chain up */
+  G_APPLICATION_CLASS (hwangsae_agent_parent_class)->dbus_unregister (app,
+      connection, object_path);
+}
+
+gboolean
+hwangsae_agent_edge_interface_handle_start (Hwangsae1DBusEdgeInterface * object,
+    GDBusMethodInvocation * invocation, gchar * arg_id, gpointer user_data)
+{
+  ChamgeReturn ret = CHAMGE_RETURN_OK;
+  g_autofree gchar *cmd = NULL;
+  g_autofree gchar *response = NULL;
+  GError *error = NULL;
+  HwangsaeAgent *self = (HwangsaeAgent *) user_data;
+  gchar *uid = NULL;
+
+  cmd =
+      g_strdup_printf ("{\"to\":\"%s\",\"method\":\"streamingStart\"}", arg_id);
```

Comment:
I think streamingStart message is missing "uri" parameter here.

Reply:
You are correct. If this first PR is OK I will add the hardcoded parameter.

Reply:
> You are correct. If this first PR is OK I will add the hardcoded parameter.

Or, I guess _Gaeul_ uses its default relay uri if the parameter is missing. As the first PR, I think this is okay as it is. My comment here is just a reminder.